text/3668-async-closure.md

```diff
@@ -0,0 +1,713 @@
+
+However, users are quickly met with the limitation that they cannot use any of the closure's captures by reference in the async block.  E.g.:
+
+```rust
+async fn parent_country(city_name: &str) -> String { todo!() }
+
+async fn main() {
+    // Collect the country names of each city in our list.
+    let mut countries = vec![];
+    for_each_city(|city_name| async {
+        countries.push(parent_country(city_name).await);
+    })
+}
+```
+
+```
+error: captured variable cannot escape `FnMut` closure body
+  --> src/main.rs
+   |
+   |       let mut countries = vec![];
+   |           ------------- variable defined here
+   |       for_each_city(|city_name| async {
+   |  _____________________________-_^
+   | |                             |
+   | |                             inferred to be a `FnMut` closure
+   | |         countries.push(parent_country(city_name).await);
+   | |         --------- variable captured here
+   | |     }).await;
+   | |_____^ returns an `async` block that contains a reference to a captured variable, which then escapes the closure body
+   |
+   = note: `FnMut` closures only have access to their captured variables while they are executing...
+   = note: ...therefore, they cannot allow references to captured variables to escape
+```
+
+The future that is returned by the closure cannot reference any of the captures of the closure, which is a limitation that makes `|| async {}` quite unusable today without needing to, for example, clone data and declare the async block `move`.
+
+In order for this to work, the `FnMut` trait would need to be ["lending"](https://smallcultfollowing.com/babysteps/blog/2023/05/09/giving-lending-and-async-closures/#async-closures-are-a-lending-pattern); however, there are [complications](https://hackmd.io/@compiler-errors/async-closures#Lending-closures-are-not-typically-FnOnce) with implementing general lending closures.
+
+# Guide Level Explanation
+
+Just as you can write functions which accept closures, you can write functions which accept async closures:
+
+```rust
+async fn takes_async_closure(f: impl async Fn(u64)) {
+    f(0).await;
+    f(1).await;
+}
+
+takes_async_closure(async |i| {
+    core::future::ready(i).await;
+    println!("done with {i}.");
+});
+```
+
+We recommend for users to write `async Fn()`/`async FnMut()`/`async FnOnce()` and `async ||` for async closures. This is more flexible than a closure returning a future for the reasons described elsewhere in this RFC.
+
+Async closures act similarly to closures, and can have parts of their their signatures specified:
+
+```rust
+// They can have arguments annotated with types:
+let arg = async |x: i32| { async_add(x, 1).await };
+
+// They can have their return types annotated:
+let ret = async || -> Vec<Id> { async_iterator.collect().await };
+
+// They can be higher-ranked:
+let hr = async |x: &str| { do_something(x).await };
+```
```

Comment:
Please consider adding an example here for "They can use the `move` modifier". You mention `async move` elsewhere in the RFC, but I think the possibility should appear when first defining `async` closures for users.